libraries/automaton/automaton.lua

```diff
@@ -0,0 +1,204 @@
+local items = require('items')
+local packet = require('packet')
+local player = require('player')
+local exp = require('expression')
+local resources = require('resources')
+local client = require('shared.client')
+local enumerable = require('enumerable')
+
+local command = require('core.command')
+
+local data, ftype = client.new('automaton_service')
+
+local res_items = resources.items
+local assembly_offset, attachment_offset = 0x2000, 0x2100
+
+ftype.fields.head = {
+    get = function(data)
+        return resources.items[data.head_id + assembly_offset]
+    end
+}
+
+ftype.fields.activated ={
+    get = function (data)
+        return player.pet_index ~= 0 --TODO: this does not allow swapping of attachment when your automaton is not the pet you have, which is wrong.
+                                     -- need to get the name of the pet and check if it matches automaton.name. getting the name seems best done a pet lib.
+    end
+}
+
+ftype.fields.frame = {
+     get = function(data)
+        return resources.items[data.frame_id + assembly_offset]
+    end
+}
+
+ftype.fields.attachments.type.base.fields.item = {
+    get = function(data)
+        return resources.items[data.id + attachment_offset]
+    end
+}
+
+local get_attachment_id = function(attachment)
+    local id = enumerable.wrap(items:find_ids(attachment)):first(exp.lookup(res_items):index('category'):is('Automaton'))
+    return id and id - attachment_offset or nil
+end
+
+local get_assembly_id = function(assembly)
+    local id = enumerable.wrap(items:find_ids(assembly)):first(exp.lookup(res_items):index('category'):is('Automaton'))
+    return id and id - assembly_offset or nil
+end
+
+ftype.fields.validate_head = {
+    data = get_assembly_id
```

Comment:
I think I was the one who said that the validation functions can just return the ID, but they really shouldn't. ID retrieval and validation should be separate, especially because the latter can be more efficient:
```lua
local item = items:find_ids(assembly:normalize())[1]
return item and res_items[item.id].category == 'Automaton'
```

Thinking about it, you can do the same approach for ID retrieval, you don't need the `enumerable` library, I think. There can not be multiple IDs for one name that differ in their category, right? So it should be enough to always just take the first one.

You do, however, need to normalize the name. I know I said before to normalize on the call-site, but I have since improved the normalization routine enough to just always call it inside the library. It will be a minor convenience boost for developers.

Reply:
I am not sure i understand what you mean here:
> There can not be multiple IDs for one name that differ in their category, right?

Here is an example of the attachment `Strobe` it has 2 entries in `items.lua`
```lua
    [2238] = {id=2238,en="Strobe",ja="ストロボ",enl="strobe",jal="ストロボ",category="General",flags=4,stack=12,targets=0,type=1},
    ....
    [8449] = {id=8449,en="Strobe",ja="ストロボ",enl="strobe",jal="ストロボ",category="Automaton",flags=0,stack=1,targets=0,type=0},
```

Reply:
That is exactly what I meant :) So I was wrong, there can be multiple entries with the same name, so you cannot use the code I used above, and should keep the `enumerable` code.